feat: sections support

selvalt7 opened the pull request:
This adds support for sections dashboard and addresses #1111.
Which means it fixes card size and also add card size restrictions based on the card configuration.
This "shouldn't" affect other types of dashboards.
![sections](https://github.com/user-attachments/assets/5f75d13a-1491-49ee-b681-683958b42fc3)

selvalt7 committed:
feat: sections support

selvalt7 committed:
fix: invisible graph in non `sections` type dashboards

selvalt7 committed:
fix: hide state time label when card is considered small

Nickduino commented:
Please merge this into the base branch, guys, we need it :-)

akloeckner commented:
Please feel free to give a first peer review. This one will take some time on my side, as I don't have too much spare time and #1128 is first in line. @selvalt7 is very productive. 😄 Thanks!

ildar170975 commented:
Honestly I made a quick look at this PR.
What I noticed that all changes affect only on a "sections" view (hope I am not wrong) - i.e. will not break traditional views.
Cannot test it in full scale myself - not using sections.

IMHO we can give this PR a go & test with participation of users who is using sections.
Hopefully will find more bugs)))

akloeckner commented on src/main.js:
```diff
@@ -1014,6 +1085,62 @@ class MiniGraphCard extends LitElement {
     }
   }
 
+  getLayoutOptions() {
```
See comment on `getGraphHeightSections`. If my presumption about the fixed rows is correct, we could reduce this to

```js
      return {
        grid_rows: 2,
        grid_columns: 2,
        grid_min_rows: 1 + fixedRows,
        grid_min_columns: 2,
      };
```

akloeckner commented on src/main.js:
```diff
@@ -127,6 +135,68 @@ class MiniGraphCard extends LitElement {
+  }
+
+  getLayoutSize(layout) {
+    if (!this.sections) {
```
Reduce? 

```js
return this.sections && layout.grid_rows === 2 ? 'small' : '';
```

And if that is enough, maybe remove the function and do the logic in the render directly... (Because it seems used only once.)

akloeckner commented on src/style.js:
```diff
@@ -12,6 +12,10 @@ const style = css`
     position: relative;
     overflow: hidden;
   }
+  ha-card.sections {
```
Is it not sufficient to set `height`?

selvalt7 replied:
Yeah height is sufficient as seen in #1155

akloeckner commented on src/style.js:
```diff
@@ -49,6 +53,9 @@ const style = css`
     order: -1;
     padding: 0 16px 8px 16px;
   }
+  ha-card[fill].sections .graph__legend {
```
Why do the margins and paddings need tweaking? If it's kind of a refactor but yields the same layout, let's consider to change this also in non-section view to keep consistency.

akloeckner commented on src/style.js:
```diff
@@ -238,6 +254,10 @@ const style = css`
     margin-top: auto;
     width: 100%;
   }
+  .sections .graph {
```
Here, too: if this is useful also in non-section views, let's consider to just add it to the regular styles.

selvalt7 replied:
I think this is only useful for section view. It makes graph position itself so that it's not clipping out of bounds.

akloeckner commented on src/main.js:
```diff
@@ -127,6 +135,68 @@ class MiniGraphCard extends LitElement {
+    return false;
+  }
+
+  getCurrentLayout() {
```
I wonder: isn't this taken care of by home assistant? If we report a minimum row count via `getLayoutOptions`, shouldn't this be reflected in `this.config.layout_options`?

See also here, maybe: https://github.com/home-assistant/frontend/blob/67217b9dd0f5ad43aafb3bba2fe2821407b25243/src/panels/lovelace/editor/card-editor/hui-card-layout-editor.ts#L68

selvalt7 replied:
No, it only clamps the values for setting the card size and for the visual editor, values set by user in `this.config.layout_options` stays untouched even when they're lower than minimum set in `getLayoutOptions`.

akloeckner commented on src/main.js:
```diff
@@ -509,7 +580,7 @@ class MiniGraphCard extends LitElement {
     const { height } = this.config;
     return svg`
-      <svg width='100%' height=${height !== 0 ? '100%' : 0} viewBox='0 0 500 ${height}'
+      <svg width='100%' height=${height !== 0 || this.sections ? '100%' : 0} viewBox='0 0 ${this.graphWidth} ${this.graphHeight}'
```
Looking at this... Why would we ever set height to zero? If this is some kind of broken remnant, we could as well remove it. It might make our life easier with min and max height in the styles. But I do not fully see the implications yet.

selvalt7 replied:
It seems this is the old way of the hidding graph part.

I don't see a reason to keep this since we have a option to hide it through `show.graph`.

I kept it there because I did not want to modify cards behaviour on dashboards other than `sections`.

akloeckner replied:
OK. Let's stick to adding sections. We can still remove this in a seperate effort.

akloeckner reviewed (commented):
As the discussion came here before I could look at the UI PR again, I used some travel time to review. I must admit, I don't use sections, so this is from a pure code review perspective now.

Forgive me if I didn't get some parts of the code completely. All suggestions are open to discussion.

selvalt7 commented on src/main.js:
```diff
@@ -127,6 +135,68 @@ class MiniGraphCard extends LitElement {
   }
 
+  checkSections() {
+    let root = document.querySelector('home-assistant');
```
`this.layout` does not seem to be defined when the graphs are created in `getConfig`. I could try to modify graphs height in `render` as it seems that `this.layout` is defined there.

As for the `this.options.layout_options` there could be a situation where user copies card to a different dashboard that is not necessarily a `sections` but maybe I am overthinking.

selvalt7 committed:
Cleaner minimum row calculation
Card is considered small when its row size is 3 or less
Simpler sections checking
Simpler card height style

jnnck commented:
Hi Guys! Thanks for your work, just wanted to let you guys know that as a sections user, i tried out this PR and it did fix my dashboards

Before:
<img width="1404" alt="Screenshot 2024-10-08 at 15 11 47" src="https://github.com/user-attachments/assets/7ab4acc6-4711-4c3d-a7f3-204b6d4f4947">

After:
<img width="1410" alt="Screenshot 2024-10-08 at 15 10 48" src="https://github.com/user-attachments/assets/0f89eb21-2775-4785-8cbc-fc75eee62767">

Personally i did like the thinner lines, but thats just preference i guess!

Nickduino commented:
Awesome. Could you show us how the labels look like on your humidity graph with the PR?

![image](https://github.com/user-attachments/assets/cd7880d9-9729-44d8-94ec-f8b9a8d9a1c7)

selvalt7 commented:
I haven't touched the labels, but I think it could be solved by changing `flex-direction` in the labels container to `column-reverse` and changing their arrangement in HTML so they appear correctly.

This would introduce some overlaps, of course.

Before:
![brave_SqDfzx3I9s](https://github.com/user-attachments/assets/f9733a89-013f-440f-871a-8e020bfb68bd)
After:
![brave_H3SzSNnlQB](https://github.com/user-attachments/assets/fe87a795-966b-4605-a9ef-94acbf98e1de)

ildar170975 commented:
Hope this PR will only touch sections and will not affect other layouts. 
I would suggest to reduce it to as minimal as possible to simplify testing and probably to review necessary changes.

Nickduino commented:
> After: ![brave_H3SzSNnlQB](https://private-user-images.githubusercontent.com/52136675/374597198-fe87a795-966b-4605-a9ef-94acbf98e1de.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgzOTc3OTIsIm5iZiI6MTcyODM5NzQ5MiwicGF0aCI6Ii81MjEzNjY3NS8zNzQ1OTcxOTgtZmU4N2E3OTUtOTY2Yi00NjA1LWE5ZWYtOTRhY2JmOThlMWRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA4VDE0MjQ1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUzYTViNGZlM2I1NWY1MzJhNzAzMTQ5MDQxNjgyZjcxYjkwMWI2ZGQ3ZTY5MzE1OGE2MGYwNmZjYzhiOGY5YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DUtTYukojmszSnaZpvu1W2qW25Ne1bWSnO2JDP-oCrg)

If you want to go that route, I think you shouldn't leave such an important gap at the bottom **and** the state should remain on top of the labels

jnnck commented:
> @jnnck Awesome. Could you show us how the labels look like on your humidity graph with the PR?

Labels before:
<img width="448" alt="Screenshot 2024-10-08 at 18 54 41" src="https://github.com/user-attachments/assets/a20fbe97-7e5c-4cda-8c71-8ec08bbf7c01">
<img width="455" alt="Screenshot 2024-10-08 at 18 54 47" src="https://github.com/user-attachments/assets/bb532342-5f23-4a5f-8e89-9bfa16988541">

Labels after:
<img width="459" alt="Screenshot 2024-10-08 at 18 53 29" src="https://github.com/user-attachments/assets/267ee8ec-ceb1-46d1-bbd0-e79672f41528">
<img width="453" alt="Screenshot 2024-10-08 at 18 53 38" src="https://github.com/user-attachments/assets/7201b9ff-6c04-4598-bb9f-7a30aa77a444">

Nickduino commented:
Will you guys implement the changes in a v0.12.2-dev.3 or something?

akloeckner commented:
Just to let you know: I plan to get this feature released. However, I am currently overloaded and I would prefer to do this properly (rather than iterating the structure over multiple releases). So, I might still need some time. Thanks for bearing with me!

ildar170975 commented on src/style.js:
```diff
@@ -4,6 +4,7 @@ const style = css`
   :host {
     display: flex;
     flex-direction: column;
+    height: 100%;
```
suggest to check this in a stock horiz stack with another card.

selvalt7 committed:
Fixed incorrect card height in horizontal stack

selvalt7 committed:
Merge remote-tracking branch 'upstream/dev' into sections-support

ildar170975 commented:
@selvalt7 
I suggest to divide this PR to 2 parts:
1. Simply adapting a height to an available area.
2. Style elements dependently on number of rows & columns.

Reasons are that this PR is very large & touches many things. And I have a feeling that still we can together improve styling (my fault also - I was not able to dedicate enough time for this).

I prepared a super short PR (https://github.com/kalkih/mini-graph-card/pull/1155) to fix a height (there are similar changes in your PR).
Let's merge it first - at least users will not have issues like "my card does not occupy a whole area", "my card overlaps another card".
Then we spend more time for THIS PR (it really needs more attention).
What do you think?

@akloeckner 
What do you think?

selvalt7 commented:
`getLayoutOptions` indeed could be moved to the separate PR as its not affecting the card that much. It's only limiting size of the card based on the configuration only affecting sections dashboard.

akloeckner commented:
I am really sorry to take so long to review your PR properly, @selvalt7! The same goes for the UI configuration PR, unfortunately.

I think, it is a good idea to try and go step by step. It will also help me to dive into the individual chunks.

For what it's worth, I _have_ been working on the sections feature, based on your PR:

https://github.com/akloeckner/mini-graph-card/tree/feat/sections

@ildar170975, I think, simpliy fixing the height to 100% did break the masonry layout, because the svg does not have a default height. Here's a diff to the `dev` branch of what I changed and tested locally in all view types:

https://github.com/kalkih/mini-graph-card/compare/dev...akloeckner:mini-graph-card:feat/sections#diff-10d28e2613270db220e1a250cde735ca7d59741ce26c5dbce7e492909e243cb2

It does not actually need the `.sections` nor the `.small` classes. (Although they are still there from @selvalt7's original code.)

The next thing, I was trying to get my head around was the graph sizing. I do not like changing its size with every re-render. Actually, its appearance also changes depending on the view, if we do this. I began to get the impression that the 500px were originally just chosen as a reasonable starting point and then everything was resized using CSS. I think, we should stick to that philosophy. (Or change it for all views.)

So: Yes, let's try and make chunks.
  1. CSS changes to enable proper layout in **all** (!) common view types. Hopefully without any JS changes.
      Either with @ildar170975's very simple fix or something more comprehensive like my work-in-progress.
  2. Maybe include the `getLayoutOptions` already, even if its only usable in UI configuration. 
      But let's apply static values there for now. And I also think, we should respect, if a user sets too small a height in yaml configuration.
  4. Think about the `.small` feature separately. I really like it, but I feel it is a bit too mixed up in the sections-only view.

selvalt7 commented:
@akloeckner
I am checking your branch and I think the squishing effect can be solved for graph line with `vector-effect="non-scaling-stroke"`.
Points are not so straight forward as `vector-effect` is not sufficient.

So I thought about the solution and I came up with one:
Instead of `circle` we use `line` with `stroke-linecap='round'` and `vector-effect='non-scaling-stroke'`.
To make the points look the same, we need to use two `line` elements for every data point.

selvalt7 committed:
Alternate points rendering
Removed graph aspect ratio
Removed setting graph size

nmehlei commented:
Hey all, First of all thanks for all the amazing work that already went into this. Since sections is getting adopted more and more in HA, it makes sense to add support.
I see the last comment is frmo January, and we will have December tomorrow. Any update here? :)

ildar170975 commented:
@nmehlei 
Hi! People who are maintaining the project are currently busy, hopefully in Christmas time these things will be sorted out!
Meanwhile, you can use a card-mod solution with `ha-card {height: 100%}`.

nmehlei commented:
Interesting, I did not pick up this workaround from the discussion. This is great to highlight, thanks!
Unfortunately, it does not work in my case. Do you see where I went wrong? Expectation would be that the left tile has the same dimensions as the other tiles. Thanks!

<img width="487" height="309" alt="Screenshot 2025-11-30 at 14 46 05" src="https://github.com/user-attachments/assets/496bf864-3e88-4b92-8ba4-f83efff5aec7" />
<img width="448" height="263" alt="Screenshot 2025-11-30 at 14 45 56" src="https://github.com/user-attachments/assets/b53c9adc-8c68-4dd2-a614-5af6f960cf74" />

ildar170975 commented:
@nmehlei 
Currently I am away from a PC, can you wait for a week? Fill free to ping me in HA community.

nmehlei commented:
Hey @ildar170975, wasn't sure how to DM you directly in HA community, but would be amazing if you could help me out here :) in case others stumble upon the same issue with the workaround until this PR gets resolved.

ildar170975 commented:
@nmehlei 
I am ildar gabdullin in Community. 
Of course solutions should be posted publicly. 
Hope in 2-3 days I will be available to help you.

ildar170975 commented:
@nmehlei 
Just got to a PC.
Sorry for a misleading, you need not only to set 100% for `ha-card`, but also to leave a "default" `display`property for `:host` - as it is described here: https://github.com/kalkih/mini-graph-card/pull/1155/changes (a short PR only for a fixed card's height).
With card-mod, you need to use this:
```
card_mod:
  style: |
    ha-card {height: 100%;}
    :host {display: unset !important;}
```
Compare:
<img width="1128" height="532" alt="image" src="https://github.com/user-attachments/assets/90aa0fb1-c184-4755-8d24-7a4094365c1c" />
```
  - theme: ''
    title: test_0
    type: sections
    sections:
      - type: grid
        cards:
          - type: custom:mini-graph-card
            entities:
              - entity: sensor.system_monitor_disk_use
            grid_options:
              columns: 6
              rows: 4
          - type: alarm-panel
            states:
              - arm_home
              - arm_away
            entity: alarm_control_panel.security
            grid_options:
              columns: 6
              rows: auto
      - type: grid
        cards:
          - type: custom:mini-graph-card
            entities:
              - entity: sensor.system_monitor_disk_use
            grid_options:
              columns: 6
              rows: auto
            card_mod:
              style: |
                ha-card {height: 100%;}
                :host { display: unset !important;}
          - type: alarm-panel
            states:
              - arm_home
              - arm_away
            entity: alarm_control_panel.security
            grid_options:
              columns: 6
              rows: auto
```
Play with a `height` option of the card to set a desired graph's height.

kattcrazy commented:
> ```
> card_mod:
>   style: |
>     ha-card {height: 100%;}
>     :host {display: unset !important;}
> ```

Much appreciated! Looking forward to more development on this front

Liquidmasl commented:
This would be a huge boon. 
I am working on another custom card that somewhat relies on other cards handling the grid correctly, the mini graph card does not currently because this is missing.

Liquidmasl commented on src/main.js:
```diff
@@ -127,6 +132,31 @@ class MiniGraphCard extends LitElement {
+    return this.sections && layout.grid_rows <= 3 ? 'small' : '';
+  }
+
+  getGraphHeightSections() {
```
Is this dead code? 
I cant see any mention of it besides the definition. 

It should drive `this.graphHeight` for sections, or it should be removed. But if sections do need dynamic height, this wire-up seams to be missing

Liquidmasl replied:
this should actually be used, and not deleted. 
in section view the graph kinda breaks if `height` in yaml is set to something very small, eg. everything below 10. the scg canvas kinda bugs out or something. 

So the svg viewBox height should be based on the grid rows, instead of using the config height. This method does exactly that, it just needs to be wired it. 

So in the end this is the only thing that should be fixed IMHO

Liquidmasl commented on src/style.js:
```diff
@@ -1,16 +1,15 @@
     padding: 16px 0;
     position: relative;
     overflow: hidden;
+    height: 100%;
```
this height is set for all `ha-card`, nut just `.sections` 

shuldnt this be in the class below? In `ha-card.sections`?

Liquidmasl replied:
still works fine in masonry view. so this does not seam to be an issue. In that case just remove the empty css class and leave this as it is.

Liquidmasl commented on src/style.js:
```diff
@@ -1,16 +1,15 @@
 import { css } from 'lit-element';
 
 const style = css`
-  :host {
-    display: flex;
-    flex-direction: column;
-  }
```
was this intentional? wouldnt this possibly break the card in non section context?

Liquidmasl replied:
doesnt seam to break anything.. still.. intentional?

Liquidmasl commented on src/main.js:
```diff
@@ -438,16 +470,27 @@ class MiniGraphCard extends LitElement {
+      @mouseover=${() => this.setTooltip(i, point[3], point[V])}
+      @mouseout=${() => (this.tooltip = {})}
+    >
+      <line
+        class='line--point--border'
+        x1=${point[X]} y1=${point[Y]} x2=${point[X]} y2=${point[Y]}
+        stroke-linecap="round" stroke-width=${this.config.line_width * 2}
+        stroke=${color} vector-effect="non-scaling-stroke"
+      />
+      <line
         class='line--point'
-        ?inactive=${this.tooltip.index !== point[3]}
-        style=${`--mcg-hover: ${color};`}
-        stroke=${color}
-        fill=${color}
-        cx=${point[X]} cy=${point[Y]} r=${this.config.line_width}
-        @mouseover=${() => this.setTooltip(i, point[3], point[V])}
-        @mouseout=${() => (this.tooltip = {})}
+        style="stroke: var(--primary-background-color, white);"
+        x1=${point[X]} y1=${point[Y]} x2=${point[X]} y2=${point[Y]}
+        stroke-linecap="round" stroke-width=${this.config.line_width}
+        vector-effect="non-scaling-stroke"
       />
+    </g>
```
This has nothing todo with the section view. It has an impact on all kind of cards, so it should probably be in its own PR.

Liquidmasl replied:
that being said.. i dont see any issue with it in manual testing..

Liquidmasl commented on src/style.js:
```diff
@@ -1,16 +1,15 @@
     overflow: hidden;
+    height: 100%;
+  }
+  ha-card.sections {
```
if height is not placed into this class, it probably has no right to exist, dont need empty css classes

Liquidmasl replied:
still works fine in masonry view. so this does not seam to be an issue. In that case just remove the empty css class and leave this as it is.

Liquidmasl commented on src/main.js:
```diff
@@ -48,6 +48,9 @@ class MiniGraphCard extends LitElement {
     this._md5Config = undefined;
+    this.sections = false;
+    this.graphHeight = 100;
+    this.graphWidth = 500;
```
this width is set but never changed again if I see this correctly. Is this fine? 
In the screenshots I see graphs with different aspect ratios, and they look fine, so I guess its not an issue. 
Still leaving the question here if this was intentional. Streching with `preserveAspectRatio='none'` may not always look great?

this is a nitpick, its probably fine

Liquidmasl replied:
manual testing shows it works just fine, no weirdness, scales correctly, looks great.

Liquidmasl reviewed (changes requested):
Most issues are rather inconsequential. But I would take a second look at the `height: 100%` in the `ha-card` css class and the removed `:host` so they dont regress non section uses of this card. 

To small things! I would love to get this merged!

Liquidmasl commented:
Well damn, I saw https://github.com/kalkih/mini-graph-card/pull/1199 to late. 

Changes in that PR might be a little cleaner than in this one, and tackle the same issues. Will take a look at it.
